pyrdp/parser/rdp/ntlmssp.py

```diff
@@ -86,3 +93,118 @@ def parseNTLMSSPAuthenticate(self, data: bytes, stream: BytesIO) -> NTLMSSPAuthe
         response = ntChallengeResponse[16 :]
 
         return NTLMSSPAuthenticatePDU(user, domain, proof, response)
+
+    def parseNTLMSSPTSRequest(self, data: bytes, stream: BytesIO) -> NTLMSSPTSRequestPDU:
+        if not ber.readUniversalTag(stream, ber.Tag.BER_TAG_SEQUENCE, True):
+            raise UnknownPDUTypeError("Invalid BER tag (%d expected)" % ber.Tag.BER_TAG_SEQUENCE)
+
+        length = ber.readLength(stream)
+        if length > len(stream.getvalue()):
+            raise ParsingError("Invalid size for TSRequest (got %d, %d bytes left)" % (length, len(stream.getvalue())))
+
+        version = None
+        negoTokens = None
+
+        # [0] version
+        if not ber.readContextualTag(stream, 0, True):
+            return NTLMSSPTSRequestPDU(version, negoTokens, data)
+        version = ber.readInteger(stream)
+
+        # [1] negoTokens
+        if not ber.readContextualTag(stream, 1, True):
+            return NTLMSSPTSRequestPDU(version, negoTokens, data)
+        ber.readUniversalTag(stream, ber.Tag.BER_TAG_SEQUENCE, True)  # SEQUENCE OF NegoDataItem
+        ber.readLength(stream)
+        ber.readUniversalTag(stream, ber.Tag.BER_TAG_SEQUENCE, True)  # NegoDataItem
+        ber.readLength(stream)
+        ber.readContextualTag(stream, 0, True)
+
+        negoTokens = BytesIO(ber.readOctetString(stream))            # NegoData
+        return NTLMSSPTSRequestPDU(version, negoTokens)
+
+    def parseNTLMSSPChallengePayload(self, data: bytes, stream: BytesIO, workstationLen: int) -> NTLMSSPChallengePayloadPDU:
+        workstation = stream.read(workstationLen)
+        return NTLMSSPChallengePayloadPDU(workstation)
+
+    def writeNTLMSSPChallenge(self, workstation: str, serverChallenge: bytes) -> bytes:
+        stream = BytesIO()
+        substream = BytesIO()
+
+        workstation = workstation.encode('utf-16le')
```

Comment:
We have pack methods for that. It should look like, check `parser/rdp/client_info.py` for inspiration:
```
        Uint16LE.pack(len(extraInfo.clientAddress), stream)
        stream.write(extraInfo.clientAddress)
```

Reply:
Forget it, this is not related to writing to the stream so encoding here is fine as a prior step.